stream.go

```diff
@@ -219,18 +222,25 @@ START:
 
 WAIT:
 	var timeout <-chan time.Time
+	var timer *time.Timer
 	writeDeadline := s.writeDeadline.Load().(time.Time)
 	if !writeDeadline.IsZero() {
 		delay := writeDeadline.Sub(time.Now())
+		timer = time.NewTimer(delay)
 		timeout = time.After(delay)
 	}
 	select {
+	case <-s.session.shutdownCh:
 	case <-s.sendNotifyCh:
-		goto START
 	case <-timeout:
 		return 0, ErrTimeout
 	}
-	return 0, nil
+	if timer != nil {
+		if !timer.Stop() {
+			<-timeout
+		}
+	}
```

Comment:
I don't see where `timer` is ever actually used as a timer? It seems to only be used in the `if timer != nil` check which makes it equivalent to `if !writeDeadline.IsZero()` or `if timeout == nil` (although comparing chan to nil.

It's not clear to me why we want to wait on `<-timeout` anyway. It seems like that will needlessly delay handling shutdown and notify updates.

Perhaps writes should be treated distinctly from reads and exit immediately on `<-s.session.shutdownCh`? I don't think the `miss a final read on shutdown` concern applies to writes, but I could be wrong!

I've never looked at this code, and you may be the first person to study it in depth in a decade! If you're making other changes sprinkling in some comments would be most welcome!

Reply:
I just noticed above in the `Read` code that `timeout` isn't a `time.After` but `timer`'s `C` chan! That makes your choice here make more sense, but I still think my comment applies: since in this Write path the timer and timeout are unrelated I think this could cause unnecessary blocking on shutdown and send notification.

Reply:
`It's not clear to me why we want to wait on <-timeout anyway. It seems like that will needlessly delay handling shutdown and notify updates.`

`t.C` will always have a value to drain if the conditional evaluates to true, so it will _never_ end up waiting.

The changes were made to `Write` in order to unify the changes with the `Read` code path, and because it represents a (tiny) memory leak.

https://pkg.go.dev/time#Timer.Stop

> Stop prevents the Timer from firing. It returns true if the call stops the timer, false if the timer has already expired or been stopped. Stop does not close the channel, to prevent a read from the channel succeeding incorrectly.
> 
> To ensure the channel is empty after a call to Stop, check the return value and drain the channel. For example, assuming the program has not received from t.C already:
> 
> ```go
> if !t.Stop() {
> 	<-t.C
> }
> ```

> func After(d [Duration](https://pkg.go.dev/time#Duration)) <-chan [Time](https://pkg.go.dev/time#Time)
> After waits for the duration to elapse and then sends the current time on the returned channel. It is equivalent to NewTimer(d).C. The underlying Timer is not recovered by the garbage collector until the timer fires. If efficiency is a concern, use NewTimer instead and call Timer.Stop if the timer is no longer needed.

Reply:
You could hit a timeout calling `Write` on `Stream A` if there's `Stream B` on the same `Session` doing a `Write` that takes longer than `Stream A`'s timeout period.

The timeout code/functionality exists already, I just added the bit to drain the channel if the timeout doesn't fire.

Reply:
And the timeout gets set via `SetWriteDeadline`

Reply:
> t.C will always have a value to drain if the conditional evaluates to true, so it will never end up waiting.

`timeout` is not `timer.C` though. They're different chans.

Reply:
🤦🏻‍♂️ 

I see the error. I missed a step. Thanks for catching it.

Reply:
Fixed with `git rebase -i`, edited the commit in the middle: https://github.com/hashicorp/yamux/pull/127/commits/6d3eff9fc40a47d08c3c389c6672ec8ad6f63d70

Reply:
Oh, and it has to go to the top after `<-s.session.shutdownCh` as the returned errors are different depending on the state of the session/how it failed. Connection reset vs Stream closed.